### ISSUE TITLE
Next version 4.x

### ISSUE THREAD

**Issue by feuzeu** [jaxon-php/jaxon-js#22]
Hi all,

Following the most recent issues on this package, #21, and #20, I've started a new branch for the next version, 4.x.

The goal for now is to implement the same features as in v3.x, but with a more modern codebase, using the latest javascript features.

I'll be glad to have some suggestions on what can be done for this purpose.

From the two latest issues posted here, we already have these:
- Use the `fetch` api for ajax calls.
- Get rid of the `eval()` calls.
- Get rid of the deprecated or outdated javascript calls, expressions or constructions.
- Add tests.

What else can we do to improve the library?
- Use javascript modules?
- Use classes?
- Use Typescript?
- Use some third party libraries instead of implementing the features (DOM, array, string, on page ready) from scratch? Which ones?
- Provide a lightweight jQuery-like library?

Your suggestions are welcome.

Thanks,
Thierry

**Comment by feuzeu** [comment]
Hi,

I just made the first commit following the @iReserveApps comment on #21.
https://github.com/jaxon-php/jaxon-js/commit/039faae14d18d66b43d647b5c92a45288a79a65b

This commit removes calls to `eval()` in the `jaxon.cmd.dom` functions.

**Comment by feuzeu** [comment]
Hi,

I've now removed all calls to `eval()` and `new Function` in the library.

When the library needs to execute custom js code, it now creates a `script` tag that saves a function with the js code as body in a dedicated attribute, which is then used to execute the js code.
The `script` tag can then be safely removed.

For example, if we want to execute `console.log("This is a simple message")`, the library adds the following tag. 
```
<script>
  jaxon.cmd.script.context.delegateCall = () => { console.log("This is a simple message") }
</script>
```
And then calls `jaxon.cmd.script.context.delegateCall()`.
I successfully tested on Chrome and Firefox.

@iReserveApps, following your comment on jQuery in #21, I checked and it appears that the calls to jQuery from the PHP response object will use the above technique to execute jQuery code received from the server.

The question that may arise now is whether this can pose performance or security issues.

**Comment by iReserveApps** [comment]
That looks very promising. 

Its beyond my understanding of JS, but the approach makes sense. 

I've tried it by executing the example
`$response->script('console.log("This is a simple message")');`

But something did not add up. 
Get a fatal on the delegate function
`TypeError: jaxon.cmd.script.context.delegateCall is not a function
    execute jaxon.core.js:1835
    call jaxon.core.js:2301`

<img width="959" alt="2023-11-17_14-56-11" src="https://github.com/jaxon-php/jaxon-js/assets/61552383/5deb59a0-c0f9-4860-a2e2-6778c769f979">

Does this makes sense to you?

**Comment by iReserveApps** [comment]
One more test is giving more clarity. 

The step 
`jaxon.config.baseDocument.body.appendChild(scriptTag);`

Is blocked by CSP:
`Content-Security-Policy: The page’s settings blocked the loading of a resource at inline (“script-src”). [jaxon.core.js:499:39](https://local/javascript/jaxon-3.3.8/jaxon.core.js)`

**Comment by iReserveApps** [comment]
And another test, following the principle here
https://csp.withgoogle.com/docs/adopting-csp.html 
> section "Refactor calls to JS APIs incompatible with CSP"

Added this to the CSP: default-src 'self' 'nonce-rAnd0m'

And then this nonce to jaxon:
`const scriptTag = jaxon.config.baseDocument.createElement('script');
        scriptTag.nonce = 'rAnd0m';
        scriptTag.setAttribute('id', scriptTagId);`

And voila, the script executes. 
Now the next question is how to generate a real nonce, which should be page specific. 

Suggestion to put it in a similar way as to the CSRF token? 
Also server generated, page specific. So a nonce here makes sense. 
Agreed @feuzeu ?

**Comment by feuzeu** [comment]
> That looks very promising.
> 
> Its beyond my understanding of JS, but the approach makes sense.
> 
> I've tried it by executing the example `$response->script('console.log("This is a simple message")');`
> 
> But something did not add up. Get a fatal on the delegate function `TypeError: jaxon.cmd.script.context.delegateCall is not a function execute jaxon.core.js:1835 call jaxon.core.js:2301`
> 
> <img alt="2023-11-17_14-56-11" width="959" src="https://user-images.githubusercontent.com/61552383/283824565-5deb59a0-c0f9-4860-a2e2-6778c769f979.png">
> Does this makes sense to you?

I'm sorry but I can't figure out what has happened. The `jaxon.tools.dom.createFunction()` function is supposed to have defined that variable. May be you can comment the line which delete the script tag in the `jaxon.tools.dom.createFunction()` function, and then add a breakpoint in that same function.
I tested and Chrome was able to step in the function is the new script tag.

**Comment by feuzeu** [comment]
> And another test, following the principle here https://csp.withgoogle.com/docs/adopting-csp.html
> 
> > section "Refactor calls to JS APIs incompatible with CSP"
> 
> Added this to the CSP: default-src 'self' 'nonce-rAnd0m'
> 
> And then this nonce to jaxon: `const scriptTag = jaxon.config.baseDocument.createElement('script'); scriptTag.nonce = 'rAnd0m'; scriptTag.setAttribute('id', scriptTagId);`
> 
> And voila, the script executes. Now the next question is how to generate a real nonce, which should be page specific.
> 
> Suggestion to put it in a similar way as to the CSRF token? Also server generated, page specific. So a nonce here makes sense. Agreed @feuzeu ?

First of all, thanks for the report.

I think we have a clear conclusion now. It is impossible to securely execute a random string of js code in a browser, which is not so surprising, after all.

Thanks also for the tip about the nonce. I don't know how its works yet, but knowing that it can solve that security issue is really great.
As I see it for now , the PHP app could insert an optional nonce in its response, which the js library would insert in the script tag. I need to check how to do that in the PHP library.

**Comment by feuzeu** [comment]
Hi all,

I just commited a big change on the new branch.
All the "classes" have been rewritten and are now defined as IIFE (Immediately Invoked Function Expressions).
The biggest advantage is the ability to define local functions and variables, with some kind of private access. No need to use the entire path in the class hierarchy to call each function anymore.
The codebase also uses more modern js code style.

As a result, the size of the minified core library (jaxon.core.min.js) is now 23.6kb, ie an impressive 33,7% reduction from the previous 35.6kb.

Next step, we need to address the CSP security issues, and add tests to the library.

**Comment by feuzeu** [comment]
Hi all,

I've finally implemented quite a different solution to the CSP security issues.
The library can now execute some simple js expressions based on a json representation (and not the raw js source code).

The code is still simple (140 lines of code in https://github.com/jaxon-php/jaxon-js/blob/v4.x/src/cmd/json.js), it is secure since no new function can be defined (excepted event handlers which actually can only be wrappers on existing functions), and it has enough features to support the functions in the library that need to run js code.

Documentation, tests and examples are coming soon.

**Comment by feuzeu** [comment]
Hi all,
 I've moved the latest changes to the main branch. That means the first 4.x release is coming soon.

The next release will include the latest improvments on the library, while keeping compatibility with the previous 3.x releases. The minified file size is now 21.9kb, and I've also started writing tests with the Jest library (https://jestjs.io).

Unfortunately, keeping compatibility with the previous release also means that the CSP issues will not be solved in the 4.x release. I've postponed it to the 5.x release, which will feature a complete rewrite of the ajax response, both in Javascript and PHP.
So the above json.js file is moved to https://github.com/jaxon-php/jaxon-js/blob/v5.x/src/cmd/json.js.

**Comment by feuzeu** [comment]
Hi,

I've just started tagging rc of the v5, on both the PHP and Javascript libraries.

Js: https://github.com/jaxon-php/jaxon-js/tree/v5.x
PHP: https://github.com/jaxon-php/jaxon-core/tree/v5.x

I've also upgraded the exemples, which now work with the above v5 branches.

https://github.com/jaxon-php/jaxon-examples/tree/v5.x/web/exp

This version on the js package includes a rewrite of the JSON representation of commands, in addition to the following new features:
- Dialogs (modal, alerts): https://github.com/jaxon-php/jaxon-examples/tree/v5.x/web/exp/dialogs
- Confirm question: https://github.com/jaxon-php/jaxon-examples/tree/v5.x/web/exp/confirm
- Ajax call pagination: https://github.com/jaxon-php/jaxon-examples/blob/v5.x/web/exp/pagination
- The mini engine to execute json expressions we have discussed earlier, which will solve the CSP security issues by removing all dynamically generated js code.

**Comment by feuzeu** [comment]
Hi,

The version 5 of the library is almost ready to be released, after long months of dev and tests.

Finally, a command to execute some js code in the browser will look like this
```json
{
    "name": "script.exec.expr",
    "args": {
        "expr": {
            "_type": "expr",
            "calls": [
                {
                    "_type": "select",
                    "_name": "body",
                    "mode": "jq"
                },
                {
                    "_type": "method",
                    "_name": "trigger",
                    "args": [
                        "touchend"
                    ]
                }
            ]
        }
    }
}
```
The corresponding js code is
```javascript
$(body).trigger('touchend');
```

Using a json expression like the above, Jaxon will be able to execute any js expression, without eval, and without giving the opportunity to anyone to insert arbitrary js code in the page. Only existing functions can be called.

More examples of such expressions can be found in these tests: https://github.com/jaxon-php/jaxon-js/blob/main/tests/call.json.test.js.
